### DISCUSSION TITLE
How to use OneOf?

### DISCUSSION THREAD

**Discussion by DenisVoronets** [square/wire#2365] [Q&A] [answered]
Hello!
I am using your great library to work with protobuf, and I'm confused how to use OneOf in generated Kotlin classes.
For example:
1) I got .proto file with this info
```
syntax = "proto2";

message SomeProto {

oneof message {
   ClassA classA = 1;
   ClassB classB = 2;
   }
}

```
2) After build via Wire library, I got Kotlin class of this .proto file.
3) And now, after `decode()` I want to know which Class of message I received
```
val messageType = SomeProto.ADAPTER.decode(receivedByteArray)
if(messageType == ClassA) {
  // do stuff for ClassA
} else if (messageType = ClassB) {
  // do stuff for ClassB
}
```
For now I check like this:
```
if(decodedMessage.ClassA != null) {
   messageType = ClassA
} else if(decodedMessage.ClassB != null) {
   messageType = ClassB
}
```
________________________________________________________
Question is: how can I use OneOf in Kotlin to handle message Type?

**Comment by oldergod** [comment, marked as answer]
`oneof`s fields are inlined. We do a null check on each one of those fields until we find the one set. On top of it, we have a runtime check which throws if more than one field within a `oneof` is set. We don't offer ways to smartcast the set fields; that might be doable with extension functions?

**Reply by DenisVoronets** [reply]
I see, thank you!
